chrome/content/zotero/xpcom/recognizePDF.js

```diff
@@ -575,16 +588,72 @@ Zotero.RecognizePDF = new function () {
 				if (res.container) newItem.setField('bookTitle', res.container);
 				if (res.publisher) newItem.setField('publisher', res.publisher);
 			}
-			
+
 			newItem.setField('libraryCatalog', 'Zotero');
-			
+
 			await newItem.saveTx();
 			return newItem;
 		}
-		
+
 		return null;
 	}
-	
+
+	/**
+	 * Tries to extract DOI from a file name
+	 * @param {str} fileName - File name to examine
+	 * @return {str} - Found DOI or null if nothing found.
+	 */
+	function _extractDOIFromFileName(fileName){
+		// Modified the expression given at https://stackoverflow.com/a/10324802
+		let doiRe = /(10[.][0-9]{4,}(?:[.][0-9]+)*@(?:(?!\.pdf)[A-Za-z0-9-._;()/])+)/;
+		let doiRes = fileName.match(doiRe);
+		if (!doiRes) {
+			return null;
+		}
+		// Replace `@` sign with `/` (`@` sign commonly used because `/` not
+		// allowed in file names).
+		// Grab first element of array (first captured group)
+		doiString = doiRes[1].replace("@", "/");
+		Zotero.debug(`Found DOI (${doiString}) in file name`);
+		return doiString;
+	}
+
+	/**
+	 * Tries to extract ISBN from a file name
+	 * @param {str} fileName - File name to examine
+	 * @return {str} - Found ISBN or null if nothing found.
+	 */
+	function _extractISBNFromFileName(fileName){
+		// Modified the expression given at https://stackoverflow.com/a/41271701
+		// Currently only searches for 13 digit ISBNs
+		let isbnRe = /(?=(?:\D*\d){13})((?:(?:978-[01])|(?:979-8))[\d-]+)/;
+		let isbnRes = fileName.match(isbnRe);
+		if (!isbnRes) {
+			return null;
+		}
+		// Grab first element of array (first captured group)
+		isbnString = isbnRes[1];
+		Zotero.debug(`Found ISBN (${isbnString}) in file name`);
+		return isbnString;
+	}
+
+	/**
+	 * Tries to extract Year from a file name
+	 * @param {str} fileName - File name to examine
+	 * @return {str} - Found year or null if nothing found.
+	 */
+	function _extractYearFromFileName(fileName){
+		let yearRe = /(?=(?:\d){4})((?:(?:20)|(?:19))[\d]{2})/;
+		let yearRes = fileName.match(yearRe);
```

Comment:
You're going to make some historians sad here. I would think some people might have PDFs with archival documents from further back, no?

But also, 1) this will match 4 digits within a larger number and 2) it seems way too complicated. Can't it just be `\b((?:19|20)\d{2})\b`?

Reply:
Yes, that would probably be simpler. In terms of point 1, I wasn't sure what a good fix was because you could have for example "author2020.pdf" which would not have a word border at the beginning. I guess maybe something like `\D((?:19|20)\d{2})\D`?

In terms of catching older years, do you think it makes more sense to include a few more prefixes to catch (e.g. 18, 17, ...) or just look for an isolated 4 digit number?

Reply:
> I guess maybe something like \D((?:19|20)\d{2})\D?

That seems reasonable.

> In terms of catching older years, do you think it makes more sense to include a few more prefixes to catch (e.g. 18, 17, ...) or just look for an isolated 4 digit number?

It probably doesn't really matter — I don't actually have any idea if people have PDFs of archival documents with years in them, and this is a last-resort feature anyway, so if it doesn't work, that's fine. And the lower the number, the more likely there'd be false positives. Maybe we should just stick to 19 and 20.